systemd: remove bash

nikstur opened the pull request:
Enforce that the closure of systemd remains bashless.

See the individual commits for more reasoning.

Review commit-by-commit.

Depended on (already merged):
- https://github.com/NixOS/nixpkgs/pull/429797
- https://github.com/NixOS/nixpkgs/pull/429438

Part of https://github.com/NixOS/nixpkgs/issues/428908

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nikstur committed:
systemd: set debug-shell to upstream default /bin/sh

Using the ambient /bin/sh makes sense to use as the debug shell because
it's inherently something from the environment.

Some years ago this was changed to an explicit Nix Store path. However,
this doesn't really make sense and makes many things (like optionally
removing bash from the system) hard while not making anything easier or
better.

nikstur committed:
iptables: add separate lib output

This enables us to use the lib output without pulling in bash.

- Enable structuredAttrs
- Enable strictDeps
- Add bash to disallowedRequisites

nikstur committed:
linux-pam: move scripts to separate output

This enables us to use the lib and out output without pulling in bash.

- Enable structuredAttrs
- Enable strictDeps
- Add bash to disallowedRequisites

nikstur committed:
systemd: remove bashInteractive

The only thing that needed bashInteractive were bash scripts for RPM.
Remove the RPM scripts (we definitely do not need) to get rid of
bashInteractive.

nikstur committed:
libpcap: add separate lib output

This enables us to use the lib output without pulling in bash.

- Enable structuredAttrs
- Enable strictDeps
- Add bash to disallowedRequisites

nikstur committed:
lvm2: move scripts to separate output

This enables us to use the lib and bin output without pulling in bash.

- Enable structuredAttrs
- Enable strictDeps
- Add bash to disallowedRequisites

nikstur committed:
krb5: move all binaries to out

This enables us to use the lib output without pulling in bash.

- Enable structuredAttrs
- Enable strictDeps
- Add bash to disallowedRequisites

nikstur commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -466,7 +465,6 @@ stdenv.mkDerivation (finalAttrs: {
     (lib.mesonOption "version-tag" version)
     (lib.mesonOption "mode" "release")
     (lib.mesonOption "tty-gid" "3") # tty in NixOS has gid 3
-    (lib.mesonOption "debug-shell" "${bashInteractive}/bin/bash")
```
As it currently stands, /bin/sh is required for any system just like /run/current-system. Both need to be setup arbitrarily early (as in before the stage 2 systemd starts) becaues of the `system()` call from libc.

 /bin/sh on NixOS is also a fully fledged bashInteractive with tab completion and all the other bells and whistles. If you have any other shell in your PATH, you can always switch to it from the debug-shell (i.e. /bin/sh).

I think /bin/sh is very sensible. It will always be there on a normal system and if you really want a system completely without a shell, you don't have to recompile systemd to achieve it.

However, I won't die on this hill, if the consensus is that we prefer `/run/current-system/sw/bin/bash`, I'm happy to change it.

nikstur replied:
Set this explicitly to /bin/sh so that this is more clear

LordGrimmauld commented on pkgs/by-name/au/audit/package.nix:
```diff
@@ -81,6 +81,7 @@ stdenv.mkDerivation (finalAttrs: {
     # capability dropping, currently mostly for plugins as those get spawned as root
     # see auditd-plugins(5)
     "--with-libcap-ng=yes"
+    "--libexecdir=${placeholder "bin"}/libexec"
```
I proposed https://github.com/linux-audit/audit-userspace/pull/492 upstream, less of a hack than trying to install to a tmpdir here. I'll pick that into https://github.com/NixOS/nixpkgs/pull/429438 once it is merged.

nikstur replied:
See this comment: https://github.com/NixOS/nixpkgs/pull/429438#discussion_r2249157504

Rebasing when https://github.com/NixOS/nixpkgs/pull/429438 is merged.

nikstur committed:
kbd: do not compress by default

Removes the dependency on bash.

Increases the binary size by 2.3MiB. However, with the 9MiB saved
by removing the excessive dictionary from cracklib (#429797), this still
results in a net reduction of the mandatory closure size.

Once kbd allows using the compression libraries instead of shelling out
to the binaries (github.com/legionus/kbd/issues/138) we can compress by
default again.

Add the package `kbdCompressed` for users that want to load custom
compressed fonts and keymaps. Alternatively, they can still uncompress
custom fonts and keymaps themselves and then call loadkeys/setfont on
it.

nikstur committed:
systemd: add bash to disallowedRequisites

philiptaron commented:
I'm really excited for this. Anything you need testing-wise, @nikstur?

nikstur committed:
systemd: remove iptables dependency

Since https://github.com/systemd/systemd/pull/17026 (v248 in 2020),
systemd can use nftables without any new dependency!

In 259, systemd plans to remove iptables suport altogether.

arianvp commented:
Needs to be rebased due to `linux-pam` update

LordGrimmauld commented:
```
rg bash /nix/store/7hm30llifnv7pn2sa3s5kmdc5myskxf5-lvm2-static-x86_64-unknown-linux-musl-2.03.33
/nix/store/7hm30llifnv7pn2sa3s5kmdc5myskxf5-lvm2-static-x86_64-unknown-linux-musl-2.03.33/nix-support/propagated-build-inputs
1:/nix/store/y14g6gck5jppfpw3b1cihg9irn5r0zf7-libaio-static-x86_64-unknown-linux-musl-0.3.113 /nix/store/6s9c1kcqm1nka0krnda028kb69l154ra-bash-interactive-static-x86_64-unknown-linux-musl-5.3p3-dev
```

Don't ask me why, but somehow bash leaks into propagated-build-inputs on static.

alyssais commented:
Every build input becomes a propagated build input on static — is that helpful?

nikstur commented:
> Every build input becomes a propagated build input on static — is that helpful?

So this whole idea of moving scripts to a separate output doesn't work on static because bash will always be propagated because its in the buildInputs?

Is there a way to fix this (i.e. to ensure that bash isn't propagated)?

alyssais commented:
Hmm.  There are other packages that just `rm -f nix-support/propagated-build-inputs`, which really is the right thing to do in any output that doesn't contain static libraries. Maybe that's the right thing to do here (for now)?

In the long term we can probably revisit this behaviour of pkgsStatic.  We've talked before about encoding absolute paths to libraries in .pc files to create references.  If we ever end up doing that I think we can drop this propagation hack once and for all, but that's a way off for now.

LordGrimmauld commented:
imo, until we have a "good" fix (removing the wild propagation on static in favor of pkg-config tricks (https://github.com/NixOS/nixpkgs/pull/394607 ?), we should just not do the disallowed requisites check on static.

nikstur commented:
Fix is here: https://github.com/NixOS/nixpkgs/pull/435567
